---
title: "RNA-Seq Analysis"
author: "Serge Merzliakov"
date: "20/06/2021"
output: html_document
---

```{r setup, include=FALSE}

BiocManager::install(c("dplyr", "AnnotationDbi", "org.Hs.eg.db", "DESeq2", "tibble", "DT"))

library(DESeq2)
library(stringr)
library(fgsea)
library(data.table)
library(ggplot2)

options(scipen=99999)
load('data/SRP034009_rse_gene.Rdata')
rse_gene@colData$condition <- as.factor(c("Control","Control","Control","Salmonella","Salmonella","Salmonella"))
dds <- DESeqDataSet(rse_gene, design = ~ condition)

LOW_EXPRESSION_THRESHOLD <- 10
keep <- rowSums(counts(dds), na.rm = T) >= LOW_EXPRESSION_THRESHOLD
dds <- dds[keep,]
dds_df <- data.frame(counts(dds))

knitr::opts_chunk$set(echo = TRUE)
```

## Overview

An RNA-seq analysis was conducted on a sample RNA counts sample, which comared gene expression in HeLa cells infected with Salmonella Typhimurium. This strain of salmonella causes a gastroenteritis in humans. The  data was obtained from Recount2 (Accession number SRP034009)


### Process 
The analysis was performed using R, using the DEseq2 package for the Differential Gene Expression Analysis (DGE) and the fgsea package for the Gene Set Enrichment Analysis (GSEA).
TODO: experimental design

## Conclusion/Results
TODO findings


## Assess The Distribution and Suitability for DESeq2 Analysis 
DESeq2 assumes a negative normal distribution, so the data was checked to assess where this was a valid assumption. The dispersion plot shows that variances exceed the means, so that the use of the DESeq2 algorithm is appropriate.

```{r assessDist, echo=FALSE}
mean_counts <- apply(dds_df[,1:ncol(dds_df)], 1, mean)
variance_counts <- apply(dds_df[,1:ncol(dds_df)], 1, var)

assess_df <- data.frame(mean_counts, variance_counts)

ggplot(assess_df) +
  geom_point((aes(x=mean_counts, y=variance_counts))) +
  scale_y_log10() +
  scale_x_log10() +
  xlab("Mean Counts per Gene") +
  ylab("Variance per Gene") +
  annotation_logticks() +
  geom_abline(show.legend=FALSE, aes(intercept = 0, slope = 1, colour='red'))

rm(assess_df, mean_counts, variance_counts)
```


## 


## Including Plots

You can also embed plots, for example:

```{r step2, echo=FALSE}

```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.


## APPENDIX - Initial Analysis


#### TODO
source + reproducbility
https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP034009

### Version Details
```{r versionInfo}
sessionInfo()
```



