---
title: "RNA-Seq Analysis"
author: "Serge Merzliakov"
date: "26/06/2021"
output:
  html_document:
    toc: TRUE
    number_sections: TRUE
    toc_float:
      collapsed : FALSE
    toc_depth: 3
theme: readable
---

```{r setup, include=FALSE}

options(width = 110)

library(DESeq2)
library(ggplot2)
library(ggrepel)
library(plotly)
library(gridExtra)
library(BiocParallel)
library(dplyr)
library(tibble)
library(data.table)
library(stringr)
library(DT)
library(here)

P_THRESHOLD <- 0.05
DATA_DIR <- "data"
RNA_SEQ_FILE <- "data/rse_gene.Rdata"
URL_TO_FILE <- "http://duffel.rail.bio/recount/v2/SRP034009/rse_gene.Rdata"

options(scipen=99999)

# create data directory if not already present
dir.create(file.path(getwd(), DATA_DIR), showWarnings = FALSE)

# load data or fetch if run for the first time
if (! file.exists(RNA_SEQ_FILE)) {
	download.file(URL_TO_FILE, destfile=RNA_SEQ_FILE)
}

load(RNA_SEQ_FILE)

rse_gene@colData$condition <- as.factor(c("Control","Control","Control","Salmonella","Salmonella","Salmonella"))
dds <- DESeqDataSet(rse_gene, design = ~ condition)

knitr::opts_chunk$set(echo=FALSE, message=FALSE, warning=FALSE)
```

# INTRODUCTION
Intracellular infections of epithelial cells are hypothesised to produce expression of gene pathways related to immune response, infection response, and inflammation. To test this hypothesis, Salmonella Typhimurium and the Hela-229 cell line were chosen. Salmonella are intracellular bacteria, which infect and replicate within both epithelial cells and macrophages, causing acute gastroenteritis in humans. The HeLa-229 human cell line is epithelial in origin, so will constitute an acceptable host for Salmonella.

This experiment will test this hypothesis using a DESeq2 analysis, with a two-group experimental design.

# RESULTS

## Overview
An DESeq2 analysis was conducted on data comparing gene expression in HeLa-229 cells infected with Salmonella Typhimurium with a control group. The analysis was performed with the DESeq2 tool, using the R package of the same name.

### Data Set

| | |
|-----------------:|:------------------------------------------------------------------------------|
|Title:| Transcriptomic profiling of HeLa-229 cells infected with Salmonella Typhimurium |
|Accession Number:| SRP034009  [https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP034009] |
|BioProject:| PRJNA231559  [https://www.ncbi.nlm.nih.gov/bioproject/PRJNA231559] |
|File URL | [http://duffel.rail.bio/recount/v2/SRP034009/rse_gene.Rdata] |


### Experimental Design
The data set was a two-group experimental design with three replicates in each group. One group was the Salmonella infected cells, the second, the control group.

### Analysis Process
The DESeq2 R package was used to perform the Differential Gene Expression Analysis (DGE) and the clusterProfiler R package for the Gene Set Enrichment Analysis (GSEA).

* An [initial analysis](#Initial) reviewed the counts, and their distribution to confirm quality and statistical criteria were satisfied.
* A [Principle Component Analysis](#PCA) was performed for quality assessment and exploratory analysis.
* A [Volcano Plot](#volcano) was generated, highlighting significant expression of genes.
* A [HeatMap](#heatmap) was generated for the most over and under expressed genes.
* A [Gene Set Enrichment Analysis](#gsea) was performed to determine to most significantly expressed pathways.

## Principal Component Analysis {#PCA}
``` {r pcaData }

# normalize counts
dds <- estimateSizeFactors(dds)

# log transformation for large data sets
dds_vst <- vst(dds)

```
A Principal Component Analysis (PCA) was performed on the data which showed that first 2 components  explain TODO 79% of the variance, and the first four components approximately TODO 94%.

The component plots show strong differentiation on the PC1 axis for the control and condition groups, with tight clustering for the control replicates, and Salmonella replicates split into two clusters, with Salmonella Replicate 1 differing from the other 2 Salmonella replicates. The heatmap shows a different pattern of of expression for Salmonella replicate and the analysis discussed in that [section](#heatmap).

The interactions between the PC2 and PC3 are less informative, with clustering of the control replicates only. The remaining principle components provided no useful insights.

``` {r pcaPlotImp, fig.height=4, fig.width=10}

plotPCA(dds_vst)

```

## DESeq2 Analysis {#analysis}

The DESeq2 analysis was generated and a MA plot generated, which showed decreasing variability with increased counts, and an approximately equal proportion of over and under expressed genes.

``` {r generateDES, cache=TRUE}
library(EnsDb.Hsapiens.v86)

register(SnowParam(workers=2, progressbar=T))

dds <- DESeq(dds,  minReplicates=3, parallel = TRUE)
dds_res <- results(dds, parallel = TRUE)
# TODO: need to increase the 'shrinkage' here
dds_res <- lfcShrink(dds=dds, res=dds_res, type="apeglm", coef = 2, parallel = TRUE) 

plotMA(dds_res, ylim=c(-10,10))

dds_res_df<- data.frame(dds_res) %>%
  rownames_to_column(var="ensgene") %>%
  mutate(GENEID=sub('\\..+', "", ensgene))  %>% # strip off ensemble Id version
  dplyr::select(-ensgene)

ens2sym <- AnnotationDbi::select(EnsDb.Hsapiens.v86, 
                                 keys = keys(EnsDb.Hsapiens.v86),
                                 columns = c("SYMBOL"))

dds_res_df <- dds_res_df %>%
              left_join(x=dds_res_df,
                        y=ens2sym,
                        by="GENEID") %>%
              mutate(symbol=SYMBOL) %>%
              dplyr::select(-SYMBOL)

```


The table below shows all the significantly differentially expressed genes.

``` {r degTable}

# Show all DEG in a table with non-NA padj and log2FC changes
dds_res_df %>%
  dplyr::filter(!is.na(padj), !is.na(log2FoldChange)) %>%
  arrange(padj) %>%
  dplyr::select(symbol, log2FoldChange, padj) %>%
  DT::datatable()

```

## Volcano Plot {#volcano}
The top most significantly expressed genes included CCL2, IL6, CXCL2, CTSL and all of these are involved in inflammation or immune response pathways. Since they are differentially expressed, the presence of Salmonella is a possible cause. The underlying presence of HPV-18 in the cell line may also explain some of this expression[^2], but this cannot be determined with the experiment design.

Overall the plot shows conclusively that significant over and under expression occurred, with somewhat more over-expression than under-expression.


``` {r volcano, fig.height=8, fig.width=10, fig.align='center'}
library(EnhancedVolcano)

GENES_TO_LABEL <- 50  #an arbitrary choice

dds_res_volcano <- dds_res_df %>%
                arrange(padj)

volcano_plot <-EnhancedVolcano(dds_res_volcano,
                           lab=dds_res_volcano$symbol,
                           selectLab = dds_res_volcano$symbol[1:GENES_TO_LABEL],
                           x = 'log2FoldChange',
                           y = 'pvalue',
                           xlim = c(-15, 20),
                           title = 'Salmonella Infected Cell DGE',
                           pCutoff = 1E-100,
                           pointSize = 2.0,
                           labSize = 5.0)

# cannot use  ggplotly - error - Error in unique.default(x)
# cannot use grid.arrange  - errors
show(volcano_plot)
```


## HeatMap {#heatmap}
A review of the most over-expressed and under-expressed genes was found to broadly support the hypothesis. However, the value of interpreting individual gene expression levels when investigating complex pathways such as immune or infection responses was limited.

``` {r heatmapPost, fig.height=8, fig.width=8, fig.align='center'}
library(pheatmap)

TOP_N <- 20

norm_counts <- data.frame(counts(dds, normalized=T))

# based on brewer.pal(n = 7, name = "RdYlBu")
hmap_colors <- c("#275591","#91BFDB","#E0F3F8","#FFFFBF","#FEE090","#FC8D59","#D73027")

# top 20 overexpressed
gene_ids <- dds_res_df %>%
            dplyr::filter(!is.na(padj)) %>% 
            arrange(desc(log2FoldChange)) %>% 
            dplyr::filter(row_number()<= TOP_N | row_number() > n()- TOP_N) %>%
            dplyr::select(GENEID) %>%
            deframe()
         
sig_norm_counts <- norm_counts[gene_ids,]

dds_metadata <- data.frame(condition=rse_gene@colData$condition, row.names = colnames(sig_norm_counts))

heatmap_plot <- pheatmap(sig_norm_counts,
                         color=hmap_colors,
                         show_rownames=T,
                         labels_row = gene_ids,
                         labels_col = rse_gene@colData$title,
                         annotation_col=dds_metadata,
                         scale="row",
                         main= "Top 20 Over/Under Expressed Genes")
show(heatmap_plot)

```

### Salmonella Replicate 1
This replicate was significantly different from the other Salmonella replicates from the PCA analysis. Compared to the other Salmonella replicates, there appears to be a group of genes under-expressed. This includes genes S100A9, FGB, S100A8, LBP and VWA1.

Using Enrichr [^5], this group of genes relates to several pathways such as Endogenous Toll-Like Receptor signalling, involved in infection responses and IL-17 Signalling, involved in inflammation responses. These pathways were not detected in the GSEA analysis, however. Overall, the expression pattern is similar enough to the other replicates to warrant further investigation.

### Salmonella Replicates 2 and 3
These two replicates were similar in their expression of various genes, with many relating to inflammation or infection responses.


## Gene Set Enrichment Analysis {#gsea}
A Gene Set Enrichment Analysis (GSEA) was performed, in which some pathways [^4] were related to infection and immunologic specific pathways, specifically:

* GOBP_RESPONSE_TO_WOUNDING
* GOBP_REGULATION_OF_RESPONSE_TO_WOUNDING
* GOBP_INFLAMMATORY_RESPONSE
* GOBP_REGULATION_OF_WOUND_HEALING

Some pathways were suggestive of tumor expression, such as GOBP_EPITHELIAL_CELL_PROLIFERATION. The presence of these may be confounding factors related to the HeLa cell line or HPV-18.

Overall, the most significant pathways had very close enrichment scores (within 10% of each other, for the top 100 pathways), with many pathways not obviously related to the experiment. Therefore, no obvious trend was revealed from the GSEA analysis, and support for the hypothesis was weak.

### Top 5 Over and Under-Expressed Pathways

``` {r gsea, fig.height=4, fig.width=12}
library(msigdbr)
library(clusterProfiler)

TOP_PATHWAYS <- 5

# Get the gene sets and wrangle
gene_sets <- msigdbr(species = "Homo sapiens", category = "C5")
gene_sets <- gene_sets %>%
  dplyr::select(gs_name, gene_symbol)

# org.Hs.eg.db VERSION
dds_res_gsea <- dds_res_df %>%
  arrange(padj) %>%
  mutate(padj = case_when(padj == 0 ~ .Machine$double.xmin,
                          TRUE ~ padj)) %>%
  mutate(gsea_metric = -log10(padj) * sign(log2FoldChange)) %>% # TODO - use stat (wald statistic) - but lfcShrink has removed this
  dplyr::filter(!is.na(gsea_metric)) %>%   
  arrange(desc(gsea_metric))

# Get the ranked GSEA vector
ranks <- dds_res_gsea %>%
  dplyr::select(symbol, gsea_metric) %>%
  dplyr::distinct(symbol, .keep_all = TRUE) %>%
  arrange(desc(gsea_metric)) %>%
  deframe()

# run GSEA
res_gsea <- GSEA(geneList = ranks,
                eps = 0,  # P-values are less than 1e-10. Set to zero for better estimation
                TERM2GENE = gene_sets)

res_gsea_df <- as.data.frame(res_gsea)
rownames(res_gsea_df) <- NULL

res_gsea_df <- res_gsea_df %>%
            arrange(desc(NES))

# plot top and bottom 5 pathways
top_bottom_pathways <- res_gsea_df %>%
  filter(row_number() <= TOP_PATHWAYS | row_number() > n() - TOP_PATHWAYS) %>%
  dplyr::select(ID, p.adjust, NES)

g <- ggplot(top_bottom_pathways, aes(reorder(ID, NES), NES)) +
  geom_col(aes(fill=NES < 0)) +
  coord_flip() +
  labs(x="Pathway",
       y="Normalized Enrichment Score",
       title=paste0("Top ",TOP_PATHWAYS, " Pathways")) +
      theme_minimal()
show(g)

```

### Top 5 GSEA Plots {.tabset}

The peak enrichment scores for the top 5 plots were very similar, although most of the pathways did not directly relate to the hypothesis being explored.

``` {r gseaPlot, fig.height=10, fig.width=12, fig.align='center'}
library(ggpubr)

top_5_pathways <- res_gsea_df %>%
  slice_max(n = TOP_PATHWAYS, order_by=NES, with_ties=F) %>%
  dplyr::select(ID, p.adjust, NES) %>%
  arrange(desc(NES))

#plot_list <- lapply(top_5_pathways$ID, function(p) return(gseaplot(res_gsea, geneSetID = p, title = p)))
#ggarrange(plotlist=plot_list, ncol=2, nrow=3)
```


#### Plot 1
``` {r gsea1}
gseaplot(res_gsea, geneSetID = "GOBP_POSITIVE_REGULATION_OF_CELL_ADHESION", title = "GOBP_POSITIVE_REGULATION_OF_CELL_ADHESION")
```


#### Plot 2
``` {r gsea2}
gseaplot(res_gsea, geneSetID = "GOCC_CELL_SURFACE", title = "GOCC_CELL_SURFACE")
```

#### Plot 3
``` {r gsea3}
gseaplot(res_gsea, geneSetID = "GOBP_WOUND_HEALING", title = "GOBP_WOUND_HEALING")
```

#### Plot 4
``` {r gsea4}
gseaplot(res_gsea, geneSetID = "GOBP_WOUND_HEALING", title = "GOBP_WOUND_HEALING")
```

#### Plot 5
``` {r gsea5}
gseaplot(res_gsea, geneSetID = "GOBP_RESPONSE_TO_WOUNDING", title = "GOBP_RESPONSE_TO_WOUNDING")
```

### Significant Pathways
The significant pathways are shown in the table below. All have very close enrichment scores - from 1.88 for the first gene to 1.68 for the first 100 pathways. So the the most significant pathways are too similarly expressed to make meaningful distinctions between them.

``` {r gseaTable, fig.height=4, fig.width=7}

# Significant Pathways in Table
res_gsea_df %>%
  dplyr::filter(p.adjust < P_THRESHOLD) %>%
  dplyr::select(ID, p.adjust, NES)  %>%
  arrange(desc(NES)) %>%
  DT::datatable()

```

# DISCUSSION

The GSEA analysis showed several pathway expression relating to inflammation, immune and bacterial infection response, along with many other pathways with similar enrichment scores. Overall, there was very limited support for the hypothesis.

There was expression of some pathways potentially related to oncogenesis, which could possibly be related to the cell line. There is previous evidence to suggest highly aberrant expression of some pathways in this cell line [^1]. Also, the cell line contains Human Papilloma Virus 18 DNA, known to cause the majority of cervical neoplasms and also implicated in the cell line's characteristic immortality. There is evidence of HPV-18 mRNA expression in HeLa cell lines[^2]. These are all confounding factors which warrant further investigation.

The value of investigating individual gene expression with heatmaps or volcano plots for evaluating complex pathways appears limited, as gene expression and functional roles are heavily context dependent. The [GSEA analysis](#gsea) was found to be slightly more effective in discovering the cellular responses to intracellular bacterial infection.

## Future Lines of Enquiry
Given the nature of the cell-line used in this experiment, future experiments could be made with other, non-neoplastic epithelial cell lines, to try and eliminate the confounding problems of aberrant tumor cell expression and viral infection. Another dimension would be to use other intracellular bacteria to see how pathway expression differs across bacterial species. It is also unclear, at what stage of the infection the samples were taken, as this may affect gene expression, as well as the natural variation in infection life-cycles between individual cells.

From a methodological perspective, it is recommended to perform RNA-seq analyses with at least two methods, from toolsets such as DESeq2, EdgeR and Limma-Voom.

# APPENDIX

## Duplicate Reads in Data {#Initial}
A review of the DEseq data found a large number of read count duplicates (87.6% of counts has at least 1 duplicate), which gave warnings in the GSEA analysis. Duplicate reads are common in RNA-seq data sets, though the high degree of duplicate may be a cause for concern and may be caused by technical sequencing issues or a high abundance of a small number of genes. 

After consideration, no action was taken to remove the duplicate values.

The table below shows duplicates of two or more counts in the raw assay data.

``` {r dups, echo=TRUE}

# assay count duplicates
assay_df <- data.frame(assay(rse_gene))
assay_df<- assay_df %>% arrange(desc(SRR1049363))
dupsCounts <- assay_df %>% group_by(SRR1049363) %>% filter(n() > 1)
head(dupsCounts, n = 12)

duplicated_n <- nrow(dupsCounts)
duplicated_percentage <- round(duplicated_n/nrow(assay_df) * 100, 1)
print(duplicated_percentage)
```


## Analysis of Distribution {#InitialDistribution}
DESeq2 assumes a negative normal distribution, so the data was checked to assess where this was a valid assumption. The dispersion plot shows that variances exceed the means, satisfying the requirement for accepting the negative binomial distribution assumption.

```{r assessDist, fig.height=4, fig.width=6, fig.align='center'}
mean_counts <- apply(assay_df[,1:ncol(assay_df)], 1, mean)
variance_counts <- apply(assay_df[,1:ncol(assay_df)], 1, var)

ggplot(data.frame(mean_counts, variance_counts)) +
        geom_point((aes(x=mean_counts, y=variance_counts))) +
        scale_y_log10(labels = scales::scientific) +
        scale_x_log10(labels = scales::scientific) +
        xlab("Mean Counts per Gene") +
        ylab("Variance per Gene") +
        annotation_logticks()

```


## Software Version Details
The details below detail the relevant system version details used to produce this analysis.
</br>
```{r versionInfo}
sessionInfo()
```

[^1]: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3737162
[^2]: https://royalsocietypublishing.org/doi/10.1098/rsob.130119
[^3]: https://www.uniprot.org/uniprot
[^4]: http://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp
[^5]: https://maayanlab.cloud/Enrichr/


## References

